src/UsingDataHooks.php

```diff
@@ -0,0 +1,298 @@
+<?php
+
+namespace UsingData;
+
+use Config;
+use MediaWiki\Hook\BeforeParserFetchTemplateRevisionRecordHook;
+use MediaWiki\Hook\GetMagicVariableIDsHook;
+use MediaWiki\Hook\ParserFirstCallInitHook;
+use MediaWiki\Hook\ParserGetVariableValueSwitchHook;
+use MediaWiki\Linker\LinkTarget;
+use MediaWiki\Revision\RevisionRecord;
+use Parser;
+use PPFrame;
+use Title;
+
+class UsingDataHooks implements
+	ParserFirstCallInitHook,
+	GetMagicVariableIDsHook,
+	BeforeParserFetchTemplateRevisionRecordHook,
+	ParserGetVariableValueSwitchHook
+{
+	public function __construct(
+		private Config $config,
+		private array $dataFrames = [],
+		private bool $searchingForData = false
+	) {
+	}
+
+	/** @inheritDoc */
+	public function onParserFirstCallInit( $parser ) {
+		$parser->setFunctionHook( 'using', [ $this, 'usingParserFunction' ], SFH_OBJECT_ARGS );
+		$parser->setFunctionHook( 'usingarg', [ $this, 'usingArgParserFunction' ], SFH_OBJECT_ARGS );
+		$parser->setFunctionHook( 'data', [ $this, 'dataParserFunction' ], SFH_OBJECT_ARGS );
+		$parser->setFunctionHook( 'ancestorname', [ $this, 'ancestorNameFunction' ], SFH_OBJECT_ARGS | SFH_NO_HASH );
+		$parser->setHook( 'using', [ $this, 'usingTag' ] );
+	}
+
+	/** @inheritDoc */
+	public function onGetMagicVariableIDs( &$variableIDs ) {
+		$variableIDs[] = 'parentname';
+		$variableIDs[] = 'selfname';
+	}
+
+	/** Returns a UsingData frame for a given page */
+	private function getDataFrame( $sourcePage, ?Title $title, Parser &$parser, PPFrame $frame ): UsingDataPPFrameDOM {
+		if ( isset( $this->dataFrames[$sourcePage] ) ) {
+			return $this->dataFrames[$sourcePage];
+		}
+
+		$this->dataFrames[$sourcePage] = new UsingDataPPFrameDOM( $frame, $sourcePage );
+		if (
+			(
+				$sourcePage === '' ||
+				Title::castFromPageReference( $parser->getPage() )->getPrefixedText()
+			) && !$parser->getOptions()->getIsSectionPreview() ) {
+			return $this->dataFrames[$sourcePage];
+		}
+
+		[ $text, $fTitle ] = $parser->fetchTemplateAndTitle( $title );
+		if ( is_object( $fTitle ) && $fTitle->getPrefixedText() != $sourcePage ) {
+			$this->dataFrames[$fTitle->getPrefixedText()] = $this->dataFrames[$sourcePage];
+		}
+
+		global $wgHooks;
+		if ( is_string( $text ) && $text != '' ) {
+			$this->searchingForData = true;
+			$clearStateHooks = $wgHooks['ParserClearState'];
+			// Other extensions tend to assume the hook is only called by wgParser and reset internal state
+			$wgHooks['ParserClearState'] = [];
+			$subParser = clone $parser;
+			$subParser->preprocess( $text, $fTitle, clone $parser->getOptions() );
+			// We might've blocked access to templates while preprocessing; should not be cached
+			$subParser->clearState();
+			$subParser->getOutput()->setText( $parser->getOutput()->getText() );
+			$wgHooks['ParserClearState'] = empty( $wgHooks['ParserClearState'] )
+				? $clearStateHooks
+				: array_merge( $clearStateHooks, $wgHooks['ParserClearState'] );
+			$parser->mPPNodeCount += $subParser->mPPNodeCount;
```

Comment:
`mPPNodeCount` is deprecated since MW1.35 , but there is no other way of updating it
I am not an MW Parser expert so I am leaving it as is :)